### PR TITLE
Add `DefensiveTruncateFrom`

### PR DESCRIPTION
**Pull request by ggwpez**
Introduces a `defensive_truncate_from` trait function to defensively truncate if `TryFrom` failed because of length bounds. Example:  
```rust
use frame_support::{BoundedVec, traits::DefensiveTruncateFrom};
use sp_runtime::traits::ConstU32;

let unbound = vec![1, 2, 3];
let bound = BoundedVec::<u8, ConstU32<2>>::defensive_truncate_from(unbound);

assert_eq!(bound, vec![1, 2]); // Truncated to two elements.
```

cc @kianenigma hopefully we can use this in the staking code to easily remove some `expect`s :smile:.

Adds:
- `TruncateFrom` trait and implementation for `BoundedVec`+`BoundedSlice`
- `DefensiveTruncateFrom` trait
- `Debug` impl for `BoundedSlice`
- `PartialEq<&'a [T]>` impl for `BoundedSlice`

Changes:
- Error type of `TryFrom<&'a [T]>` for `BoundedSlice` from `()` to `&[T]` analogous to `BoundedVec`

### CODE REVIEW

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/testing-complex-frame-pallets-discussion-tools/356/7

**Comment by kianenigma** [comment]
Was discussing it with @Ank4n: can you add a `defensive_min()` and `defensive_max` as well? I think there are a LOT of use cases where we do these operations defensively.

File: frame/support/src/traits/misc.rs [hunk 1]
```diff
@@ -369,6 +370,43 @@ impl<T: Saturating + CheckedAdd + CheckedMul + CheckedSub> DefensiveSaturating f
...
+	/// let bound = BoundedVec::<u8, ConstU32<2>>::defensive_truncate_from(unbound);
+	///
+	/// assert_eq!(bound, vec![1, 2]);
+	/// ```
+	fn defensive_truncate_from(unbound: T) -> Self;
+}
+
+impl<T, U> DefensiveTruncateFrom<U> for T
+where
+	// NOTE: We use the fact that `BoundedVec` and
```
Comment (kianenigma) [inline frame/support/src/traits/misc.rs:393]: https://github.com/paritytech/substrate/discussions/12404 🙈

**Review by kianenigma — APPROVED** [review]
LGTM, although it is worth noting whether these defensive ops need to be in frame-support anyway.

**Comment by ggwpez** [comment]
> LGTM, although it is worth noting whether these defensive ops need to be in frame-support anyway.

The `defensive!` macro uses `frame_support::log::error`, but it seems to just be a re-export of the `log` crate, so could be moved IIUC.

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot rebase

**Comment by ggwpez** [comment]
bot rebase

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**